### PR TITLE
Transform new converter call to arraytranslate node

### PR DESCRIPTION
**Pull request by dchopra001**
`StringCoding.implEncodeAsciiArray` is nearly identical in implementation and use case to the old `US_ASCII/Encoder.encodeASCII` method. This PR adds support in VP for transforming `StringCoding.implEncodeAsciiArray` into an `arraytranslate` that can then be take advantage of pre-existing acceleration for this workload.

### CODE REVIEW

**Comment by dchopra001** [comment]
Tagging @r30shah @hzongaro @vijaysun-omr for review.

**Review by hzongaro — COMMENTED** [review]
I think the changes look good, though I will defer to @r30shah for a more thorough review.

May I ask you to add a little more detail to the commit message?  Other than that suggestion, I have just a couple of cosmetic comments.

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 1]
```diff
@@ -2663,25 +2666,25 @@ TR::TreeTop* OMR::ValuePropagation::createConverterCallNodeAfterStores(
    if (lenRef)
       len = TR::Node::createLoad(root, lenRef);
    else
-      len = isISO88591Encoder?  root->getChild(childId+4)->duplicateTree() : root->getChild(childId+2)->duplicateTree();
+      len = (isISO88591Encoder || isAsciiEncoder)?  root->getChild(childId+4)->duplicateTree() : root->getChild(childId+2)->duplicateTree();
```
Comment (r30shah) [inline compiler/optimizer/ValuePropagationCommon.cpp:2669]: Please insert spaces at appropriate place.

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 2]
```diff
@@ -2663,25 +2666,25 @@ TR::TreeTop* OMR::ValuePropagation::createConverterCallNodeAfterStores(
...
 
-   isISO88591Encoder? root->setAndIncChild(childId+4, len): root->setAndIncChild(childId+2, len);
+   (isISO88591Encoder || isAsciiEncoder)? root->setAndIncChild(childId+4, len): root->setAndIncChild(childId+2, len);
```
Comment (r30shah) [inline compiler/optimizer/ValuePropagationCommon.cpp:2671]: Same as above

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 3]
```diff
@@ -2663,25 +2666,25 @@ TR::TreeTop* OMR::ValuePropagation::createConverterCallNodeAfterStores(
...
 
 
    if (dstRef)
       dst = TR::Node::createLoad(root, dstRef);
    else
-      dst = isISO88591Encoder ? root->getChild(childId+2)->duplicateTree(): root->getChild(childId+3)->duplicateTree();
+      dst = (isISO88591Encoder || isAsciiEncoder) ? root->getChild(childId+2)->duplicateTree(): root->getChild(childId+3)->duplicateTree();
 
-   isISO88591Encoder?root->setAndIncChild(childId+2, dst): root->setAndIncChild(childId+3, dst);
+   (isISO88591Encoder || isAsciiEncoder)?root->setAndIncChild(childId+2, dst): root->setAndIncChild(childId+3, dst);
```
Comment (r30shah) [inline compiler/optimizer/ValuePropagationCommon.cpp:2679]: Please insert appropriate spaces.

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 4]
```diff
@@ -2663,25 +2666,25 @@ TR::TreeTop* OMR::ValuePropagation::createConverterCallNodeAfterStores(
...
 
 
    if (dstOffRef)
       dstOff = TR::Node::createLoad(root, dstOffRef);
    else
-      dstOff = isISO88591Encoder ? root->getChild(childId+3)->duplicateTree(): root->getChild(childId+4)->duplicateTree();
+      dstOff = (isISO88591Encoder || isAsciiEncoder) ? root->getChild(childId+3)->duplicateTree(): root->getChild(childId+4)->duplicateTree();
 
-   isISO88591Encoder?root->setAndIncChild(childId+3, dstOff): root->setAndIncChild(childId+4, dstOff);
+   (isISO88591Encoder || isAsciiEncoder)?root->setAndIncChild(childId+3, dstOff): root->setAndIncChild(childId+4, dstOff);
```
Comment (r30shah) [inline compiler/optimizer/ValuePropagationCommon.cpp:2687]: Same as above

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 5]
```diff
@@ -2663,25 +2666,25 @@ TR::TreeTop* OMR::ValuePropagation::createConverterCallNodeAfterStores(
...
 
-   isISO88591Encoder?root->setAndIncChild(childId+2, dst): root->setAndIncChild(childId+3, dst);
+   (isISO88591Encoder || isAsciiEncoder)?root->setAndIncChild(childId+2, dst): root->setAndIncChild(childId+3, dst);
 
 
    if (dstOffRef)
       dstOff = TR::Node::createLoad(root, dstOffRef);
    else
-      dstOff = isISO88591Encoder ? root->getChild(childId+3)->duplicateTree(): root->getChild(childId+4)->duplicateTree();
+      dstOff = (isISO88591Encoder || isAsciiEncoder) ? root->getChild(childId+3)->duplicateTree(): root->getChild(childId+4)->duplicateTree();
```
Comment (r30shah) [inline compiler/optimizer/ValuePropagationCommon.cpp:2685]: Same as above

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 6]
```diff
@@ -2371,7 +2372,7 @@ void OMR::ValuePropagation::generateArrayTranslateNode(TR::TreeTop *callTree,TR:
    TR::Node *srcOff = srcOffNode->createLongIfNeeded();
    TR::Node *lenNode = NULL;
    TR::Node *len = NULL;
-   if (!isISO88591Encoder)
+   if (!(isISO88591Encoder || isAsciiEncoder))
```
Comment (r30shah) [inline compiler/optimizer/ValuePropagationCommon.cpp:2375]: Just a suggestion, do you think this has better readability compared what you have now? 
```suggestion
   if (!isISO88591Encoder  && !isAsciiEncoder)
```

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 7]
```diff
@@ -4209,13 +4215,13 @@ void OMR::ValuePropagation::transformConverterCall(TR::TreeTop *callTree)
    srcOff = callNode->getChild(childId++);//->createLongIfNeeded();
 
 
-   if (!isISO88591Encoder)
+   if (!(isISO88591Encoder || isAsciiEncoder))
```
Comment (r30shah) [inline compiler/optimizer/ValuePropagationCommon.cpp:4218]: Similar suggestion as given in line 2375

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 8]
```diff
@@ -4338,6 +4344,7 @@ void OMR::ValuePropagation::transformConverterCall(TR::TreeTop *callTree)
 
          case TR::sun_nio_cs_US_ASCII_Encoder_encodeASCII: threshold = 0; break;
          case TR::sun_nio_cs_US_ASCII_Decoder_decodeASCII: threshold = 0; break;
+         case TR::java_lang_StringCoding_implEncodeAsciiArray: threshold = 0; break;
```
Comment (r30shah) [inline compiler/optimizer/ValuePropagationCommon.cpp:4347]: Something that already existed in the code. I assume original intention was to have different threshold for different methods, but looking at the values set, should we refactor this to have methods that needs threshold 0 combined ?

**Review by r30shah — CHANGES_REQUESTED** [review]
Some minor nitpicks, but overall it looks good to me.

**Comment by vijaysun-omr** [comment]
@zl-wang and @knn-k / @Akira1Saitoh fyi

**Comment by knn-k** [comment]
AArch64 codegen does not implement arraytranslate now.  Issue #6092.

File: compiler/optimizer/ValuePropagationCommon.cpp (hunk 6 above)
Comment (dchopra001) [reply compiler/optimizer/ValuePropagationCommon.cpp:2375]: I personally don't have a preference in this scenario. I previously used to write them the way you suggested above. 

However in my previous PRs reviewers have suggested adding `()` around negation in `if` statements. I think the idea there was that in `if (!isISO88591Encoder)` it's easy to miss the `!` for someone skimming through the code.

If you prefer your approach, then I'll revert back to that. Let me know.

Comment (r30shah) [reply compiler/optimizer/ValuePropagationCommon.cpp:2375]: Either way works. Just had a thought, you can keep the code as it is. But please add spaces at appropriate places.

File: compiler/optimizer/ValuePropagationCommon.cpp (hunk 8 above)
Comment (dchopra001) [reply compiler/optimizer/ValuePropagationCommon.cpp:4347]: How about the following reduction?
```
int32_t threshold = 0;
if (rm == TR::sun_nio_cs_ext_SBCS_Encoder_encodeSBCS)
   {
   threshold = 11;
   }
```
We can put the names of all converters (and threshold result) in a comment in order to preserve the prior analysis done on these methods.

Comment (dchopra001) [reply compiler/optimizer/ValuePropagationCommon.cpp]: @hzongaro @r30shah The above refactoring and all the remaining formatting changes are available here: https://github.com/eclipse/omr/commit/08ce7c02b3e51cfc25b05e33a82bc0f4e6a4199e.

Once you both approve I'll squash the above commit.

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 9]
```diff
@@ -4332,20 +4338,29 @@ void OMR::ValuePropagation::transformConverterCall(TR::TreeTop *callTree)
...
-         default: threshold = 0; break;
+         case TR::sun_nio_cs_ISO_8859_1_Encoder_encodeISOArray:
+         case TR::java_lang_StringCoding_implEncodeISOArray:
+         case TR::sun_nio_cs_ISO_8859_1_Decoder_decodeISO8859_1:
+            threshold = 0;
+            break;
+         case TR::sun_nio_cs_US_ASCII_Encoder_encodeASCII:
+         case TR::sun_nio_cs_US_ASCII_Decoder_decodeASCII:
+         case TR::java_lang_StringCoding_implEncodeAsciiArray:
+            threshold = 0;
```
Comment (r30shah) [inline compiler/optimizer/ValuePropagationCommon.cpp:4349]: Any particular reason why we have this in group of 3? Just a question (given the previous state this looks better), but I suggested to bring all methods that has threshold 0 together.

**Review by r30shah — APPROVED** [review]
Changes looks good to me. Just a minor question. No changes needed in the code.

File: compiler/optimizer/ValuePropagationCommon.cpp (hunk 9 above)
Comment (dchopra001) [reply compiler/optimizer/ValuePropagationCommon.cpp:4349]: I misread your suggestion and thought you wanted it to be separated like this as well. It does keep similar named/functioning methods together, but you're right in that it's still got some redundancy. I can change it again if you prefer your original suggestion. Let me know. Thanks!

**Review by hzongaro — APPROVED** [review]
Looks good.  Thanks!

File: compiler/optimizer/ValuePropagationCommon.cpp (hunk 8 above)
Comment (hzongaro) [reply compiler/optimizer/ValuePropagationCommon.cpp]: > Once you both approve I'll squash the above commit.

Please go ahead and squash.  Thanks!

File: compiler/optimizer/ValuePropagationCommon.cpp (hunk 9 above)
Comment (dchopra001) [reply compiler/optimizer/ValuePropagationCommon.cpp]: I've grouped them all together with the default case and added a comment to explain why the names are still present in the codebase: https://github.com/eclipse/omr/commit/81f30e4f8eaafb7905c39ee02214ded71d3be315

File: compiler/optimizer/ValuePropagationCommon.cpp (hunk 8 above)
Comment (dchopra001) [reply compiler/optimizer/ValuePropagationCommon.cpp]: Commits are squashed and I've added an environment option to selectively disable this transformation (for easier defect management/problem isolation). I'll remove the option once the code is baked into the builds for a few months.

**Review by r30shah — APPROVED** [review]
LGTM

**Review by hzongaro — COMMENTED** [review]
I had a question about the use of the environment variable to control this transformation.

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 10]
```diff
@@ -4198,6 +4205,7 @@ void OMR::ValuePropagation::transformConverterCall(TR::TreeTop *callTree)
    TR_ResolvedMethod *m = symbol->getResolvedMethodSymbol()->getResolvedMethod();
    bool isISO88591Encoder = (rm == TR::sun_nio_cs_ISO_8859_1_Encoder_encodeISOArray
                              || rm == TR::java_lang_StringCoding_implEncodeISOArray);
+   bool isAsciiEncoder = (rm == TR::java_lang_StringCoding_implEncodeAsciiArray);
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply compiler/optimizer/ValuePropagationCommon.cpp]: You're right. Fixed here: https://github.com/eclipse/omr/commit/7c2e3e1076704034411553fe3f17e7fd70956e1e

**Review by hzongaro — APPROVED** [review]
Looks good.  Thanks!

**Comment by hzongaro** [comment]
Jenkins build all

**Comment by hzongaro** [comment]
[x86-64 macOS test failure](7247/checks?check_run_id=21616629563) is due to known issue #7181

**Comment by hzongaro** [comment]
OMR testing was successful, other than failures due to known issues.  Downstream OpenJ9 testing with this change was also successful.  Merging.

**Pull request merged**